### PR TITLE
Bug 1771

### PR DESCRIPTION
**Pull request by hippo91**
This PR fixes the issue #1771.
I added a test when a ``Call`` node is encountered. This test calls a method that returns True if the ``FunctionDef`` (inferred from the ``Call`` node) corresponds to a function that will never returns.

### CODE REVIEW

**Comment by hippo91** [comment]
@PCManticore what do you mean by dehardcoding `sys.exit` and what kind of customization do you think of?

**Comment by PCManticore** [comment]
This needs to be customized, so it can be a comma separated value that you can check against. Here's one example: https://github.com/PyCQA/pylint/blob/928e16431134a367a78b63540b7cbe5dd9967539/pylint/checkers/typecheck.py#L526

**Comment by PCManticore** [comment]
sys.exit is still going to be hardcoded, but in a config value :)

**Comment by hippo91** [comment]
OK i see. And do i put those customizations inside ``refactoring.py`` or somewhere else?
Thanks @PCManticore !

**Comment by PCManticore** [comment]
Yep, sounds good to me to have them in `refactoring`

File: pylint/checkers/refactoring.py [hunk 1]
```diff
@@ -29,11 +29,33 @@
 from pylint import utils as lint_utils
 from pylint.checkers import utils
 
+# sys.exit method never returns
+NEVER_RETURNING_FUNCTIONS = ('sys.exit',)
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: @hippo91 this nees to be a comma-separated tuple that is extensible by users, here's an example https://github.com/PyCQA/pylint/blob/928e16431134a367a78b63540b7cbe5dd9967539/pylint/checkers/typecheck.py#L526. This way we allow the users to provide any functions they want

File: pylint/checkers/refactoring.py [hunk 2]
```diff
@@ -29,11 +29,33 @@
...
+        never_returning_functions (tuple): complete names of the functions
+                                                         that never returns
+
+    Returns:
+        bool: True if the function never returns, False otherwise.
+    """
+    if never_returning_functions is None:
+        never_returning_functions = NEVER_RETURNING_FUNCTIONS
+    try:
+        complete_func_name = ".".join((node.root().name, node.name))
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: Wouldn't `.qname()` work as well?

File: pylint/checkers/refactoring.py [hunk 3]
```diff
@@ -29,11 +29,33 @@
...
+        if complete_func_name in never_returning_functions:
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: You could make the list of functions a set so that we have instant lookup

**Comment by hippo91** [comment]
@PCManticore could you check the options of ``RefactoringChecker`` to be sure it is what you wanted.
It's the first time i use those options. Thanks!

File: pylint/checkers/refactoring.py [hunk 4]
```diff
@@ -567,10 +525,79 @@ def _check_consistent_returns(self, node):
...
+        """Return True if the function never returns. False otherwise.
+
+        Args:
+            node (astroid.FunctionDef): function definition node to be analyzed.
+
+        Returns:
+            bool: True if the function never returns, False otherwise.
+        """
+        try:
+            if node.qname() in self._never_returning_functions:
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: `return node.qname() in ...`

File: pylint/test/functional/inconsistent_returns.py [hunk 5]
```diff
@@ -158,6 +165,11 @@ def not_consistent_returns_inner(arg): # [inconsistent-return-statements]
                 return arg
     return not_consistent_returns_inner
 
+def bug_1771_counter_example(var): # [inconsistent-return-statements]
```
Comment (PCManticore) [inline pylint/test/functional/inconsistent_returns.py]: Can you also add a test where the never returning function is customized through a rcfile? Check in `functional` directory, we have a couple of those where the option is defined in a `.rc`file

File: pylint/checkers/refactoring.py [hunk 6]
```diff
@@ -179,6 +132,10 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                 {'default': 5, 'type': 'int', 'metavar': '<int>',
                  'help': 'Maximum number of nested blocks for function / '
                          'method body'}
+               ),
+               ('never-returning-functions',
+                {'default': ('optparse.Values', 'sys.exit',),
+                 'help': 'Complete name of functions that never returns'}
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: We might explain in the help string for what are these

File: pylint/checkers/refactoring.py [hunk 7]
```diff
@@ -179,6 +132,10 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                 {'default': 5, 'type': 'int', 'metavar': '<int>',
                  'help': 'Maximum number of nested blocks for function / '
                          'method body'}
+               ),
+               ('never-returning-functions',
+                {'default': ('optparse.Values', 'sys.exit',),
```
Comment (PCManticore) [inline pylint/checkers/refactoring.py]: I am not sure, but I think you need to pass the `type` option as `csv`

**Comment by hippo91** [comment]
You'are welcome @The-Compiler !

**Pull request merged**